### PR TITLE
WiFi backend

### LINKED ISSUE #1367: Implement RobotCommunicator
### Description of the task

We need a system to dynamically connect/disconnect from robots and manage alive robots.
That system will also need to be able to send vision and primitives and receive robot status in return.

This ticket involves developing a Robot Communicator to handle the requirements outlined in the acceptance criteria.

### Acceptance criteria

- [ ] Need to send/receive protobuf reliably from multiple robots on the network
- [ ] Robots need to be able to dynamically connect/disconnect, and the communicator should keep track
- [ ] Timeouts need to be detected when a robot hasn't responded w/ a status
- [ ] Networking architecture should be implemented according to Network Architecture RFC:
https://docs.google.com/document/d/13PE2qA42idP79PFpLELDwhmxktiL3KnXso3iBd6dWQo/edit?usp=sharing
### Blocked By
N/A

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description

Implements the WifiBackend to communicate between AI and the Robots

### Testing Done
Unit tests on translation functions (except primitives)
Tested Wifi Backend manually, pending testing w/ actual robot

### Resolved Issues

resolves #1367 

### Length Justification

I can't break it any further, most of the lines are documentation. More code will be added as part of:
https://github.com/UBC-Thunderbots/Software/issues/1408
https://github.com/UBC-Thunderbots/Software/issues/1409

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/shared/proto/geometry.proto
Comment (akhilveeraghanta) [inline src/shared/proto/geometry.proto]: please ignore `.proto` changes, they will go away once #1389 gets merged

File: src/software/backend/radio_backend.h
Comment (MathewMacDougall) [inline src/software/backend/radio_backend.h]: I'm assuming this is from the branch you're based on, not your intended changes?

File: src/software/proto/message_translation/protobuf_primitive_visitor.cpp
Comment (jonathanlew) [inline src/software/proto/message_translation/protobuf_primitive_visitor.cpp]:
```suggestion
std::unique_ptr<RadioPrimitiveMsg> ProtobufPrimitiveVisitor::getRadioPrimitiveMsg(const Primitive & primitive)
```

File: src/software/backend/wifi_backend.h
Comment (MathewMacDougall) [inline src/software/backend/wifi_backend.h:13]: Should the `NetworkConfig` dynamic parameter be passed in here?

File: src/software/proto/message_translation/protobuf_primitive_visitor.cpp
Comment (jonathanlew) [inline src/software/proto/message_translation/protobuf_primitive_visitor.cpp]:
```suggestion
primitive.accept(*this);
```

File: src/software/proto/message_translation/protobuf_primitive_visitor.cpp
Comment (jonathanlew) [inline src/software/proto/message_translation/protobuf_primitive_visitor.cpp]: We can avoid this by taking a primitive as an argument. Take a look at how Navigator is done.

File: src/software/backend/wifi_backend.cpp
Comment (jonathanlew) [inline src/software/backend/wifi_backend.cpp:53]: We should standardize this with how we do unique ptrs. If we decide on the unique ptr way, then this should be changed to that.

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Good to see the progress being made here. Here are some initial comments

File: src/software/proto/message_translation/protobuf_message_translator.h
Comment (MathewMacDougall) [inline src/software/proto/message_translation/protobuf_message_translator.h]: Any reason these can't be free functions since they're just static functions of this class with no member variables.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looks sane overall, just a few small comments

File: src/shared/constants.h
Comment (kelvinkoon) [inline src/shared/constants.h]: Is there a reason why it's defined as 16x21?

**Review by kelvinkoon — COMMENTED** [review]
Looks like it's on the right track so far :+1:

**Review by garethellis0 — COMMENTED** [review]
Just taking a really quick pass, this seems sane to me. Let me know if you want a more in-depth review!

File: src/shared/proto/tbots_software_msgs.proto
Comment (akhilveeraghanta) [reply src/shared/proto/tbots_software_msgs.proto:30]: just so we get a sense of latency from when the msg was created and when it gets received on the robot

File: src/software/backend/wifi_backend.cpp
Comment (akhilveeraghanta) [reply src/software/backend/wifi_backend.cpp]: fixed

File: src/software/backend/wifi_backend.h
Comment (akhilveeraghanta) [reply src/software/backend/wifi_backend.h]: fixed

File: src/software/proto/message_translation/protobuf_primitive_visitor.cpp
Comment (akhilveeraghanta) [reply src/software/proto/message_translation/protobuf_primitive_visitor.cpp]: fixed

File: src/software/proto/message_translation/protobuf_primitive_visitor.cpp
Comment (akhilveeraghanta) [reply src/software/proto/message_translation/protobuf_primitive_visitor.cpp]: fixed

File: src/software/proto/message_translation/protobuf_primitive_visitor.cpp
Comment (akhilveeraghanta) [reply src/software/proto/message_translation/protobuf_primitive_visitor.cpp]: fixed

File: src/software/backend/radio_backend.h
Comment (akhilveeraghanta) [reply src/software/backend/radio_backend.h]: yup put it back

I think its unused though, we can probably safely remote it at some point

File: src/software/proto/message_translation/protobuf_message_translator.h
Comment (akhilveeraghanta) [reply src/software/proto/message_translation/protobuf_message_translator.h]: yup thats true, the class is unnecessary, turned it into a namespace

File: src/shared/constants.h
Comment (akhilveeraghanta) [reply src/shared/constants.h]: its the size of the channels, we have 16 groups and each group is 21 characters long

we can't use constants here instead of 16/21 because its a const char array

File: src/software/backend/wifi_backend.h
Comment (akhilveeraghanta) [reply src/software/backend/wifi_backend.h:13]: im holding off on dependency injection until we have the new system if possible

File: src/software/backend/wifi_backend.cpp
Comment (akhilveeraghanta) [inline src/software/backend/wifi_backend.cpp:21]: we should have a way to inject runtime arguments w/ the new dp system, this is the best I could do with what we have

the interface defaults to the value in the yaml file, and then the channel can change on the callback

thoughts on how we can do this better?

File: src/software/backend/wifi_backend.cpp
Comment (akhilveeraghanta) [inline src/software/backend/wifi_backend.cpp]: we should have a way to inject runtime arguments w/ the new dp system, this is the best I could do with what we have

the interface defaults to the value in the yaml file, and then the channel can change on the callback

thoughts on how we can do this better?

File: src/shared/constants.h
Comment (garethellis0) [reply src/shared/constants.h]: could we at least create these as `#define`'s?

File: src/shared/proto/tbots_software_msgs.proto
Comment (garethellis0) [reply src/shared/proto/tbots_software_msgs.proto:30]: Is the intent to eventually remove this message?

File: src/software/backend/wifi_backend.cpp
Comment (garethellis0) [reply src/software/backend/wifi_backend.cpp]: The only thing I can think of off the top of my head would be to pass the config object in instead of just using the singleton?

File: src/software/backend/wifi_backend.cpp
Comment (garethellis0) [reply src/software/backend/wifi_backend.cpp:53]: I think we ended up deciding on the unique ptr way?

File: src/software/backend/wifi_backend.cpp
Comment (garethellis0) [reply src/software/backend/wifi_backend.cpp]: Ah, just saw your comment reply to Mat.

File: src/software/backend/wifi_backend.h
Comment (garethellis0) [reply src/software/backend/wifi_backend.h:13]: Why? The interface is effectively the same no? Might as well do it the "more correct" way with the current system

File: src/software/proto/message_translation/protobuf_message_translator.h
Comment (MathewMacDougall) [reply src/software/proto/message_translation/protobuf_message_translator.h]: Do we even need the namespace? Probably worth a talk about how we want to do namespace since ages (ie. a year?) ago we sort of decided they weren't worth it since we don't need to "protect" our code or other people's code from conflicts since we "own" everything, and they _can_ cause headaches for newer people.

Obviously we aren't consistent currently and a few things still have them but worth making the call now how we want to proceed going forwards.

File: src/software/proto/message_translation/protobuf_message_translation.cpp
Comment (MathewMacDougall) [inline src/software/proto/message_translation/protobuf_message_translation.cpp]:
```suggestion
        // Since the unique_ptr immediately loses scope after the copy, the memory is
```

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
not too much to add rn

File: src/software/backend/wifi_backend.cpp
Comment (akhilveeraghanta) [reply src/software/backend/wifi_backend.cpp:53]: I tried this and its not possible to do the unique_ptr approach unfortunately, the callbacks can't be defined like this:

`std::function<void(std::unique_ptr<ReceiveProto>)> receive_callback)`

thoughts on leaving it as is?

File: src/shared/proto/tbots_software_msgs.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_software_msgs.proto:30]: What I meant was that I think that wrapping this one value in a message is unnecessary

File: src/software/backend/wifi_backend.cpp
Comment (jonathanlew) [reply src/software/backend/wifi_backend.cpp:53]: the other alternative to avoid copying values would be shared ptr but I'm not sure how ppl feel about that

**Review by jonathanlew — COMMENTED** [review]
Thanks for cleaning things up, it's looking much better

File: src/software/proto/message_translation/protobuf_message_translation.cpp
Comment (akhilveeraghanta) [reply src/software/proto/message_translation/protobuf_message_translation.cpp]: fixed

File: src/shared/constants.h
Comment (akhilveeraghanta) [reply src/shared/constants.h]: fixed

File: src/shared/proto/tbots_software_msgs.proto
Comment (akhilveeraghanta) [reply src/shared/proto/tbots_software_msgs.proto:30]: I think its worth keeping the timestamp accurate to the network time, to be sure we aren't getting worse than acceptable latency in vision msgs or primitive msgs on the robot

I put it in another msg for compatibility reasons, and so if we decide to change how we represent time (nanos) or whatever it makes it easier

similar to this style by google: https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/timestamp.proto

File: src/software/backend/wifi_backend.h
Comment (akhilveeraghanta) [reply src/software/backend/wifi_backend.h:13]: how do we want to go about this? its not super straightforward because of the generic factory. Can we dependency inject and keep the generic factory still generic?

alternatively can have a function to provide the parameters after the backend has been constructed with a `wifi_backend.storeNetworkConfig(config_here)` in `full_system` and inject it like that. 

But that adds custom checks for the backend which kind of sucks and we should avoid if possible.

Thoughts on the best way to do this?

File: src/software/backend/wifi_backend.cpp
Comment (akhilveeraghanta) [reply src/software/backend/wifi_backend.cpp:53]: This won't work with `shared_ptrs` because `shared_ptrs` don't have `.release()` IIRC

File: src/shared/proto/tbots_software_msgs.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_software_msgs.proto:30]: ok then TbotsRobotMsg would need to be changed too (maybe out of scope)

File: src/software/backend/wifi_backend.cpp
Comment (henryymliu) [reply src/software/backend/wifi_backend.cpp:53]: @akhilveeraghanta have you tried passing `unique_ptr` by reference or using `std::move`?

Comment (jonathanlew) [reply src/software/backend/wifi_backend.cpp:53]: Curious to see whether that works. It would need to be copy-constructable https://stackoverflow.com/questions/31691309/how-to-return-a-stdfunction-that-contains-a-stdunique-ptr-from-a-generalized

Comment (akhilveeraghanta) [reply src/software/backend/wifi_backend.cpp:53]: `std::move` doesn't work that was what I tried, it complains about r-value and l-value stuff when u bind the function. And I was generally avoiding passing the smart ptr by reference because that is usually frowned upon.

We are using unique_ptrs to remove raw ptrs during protobuf message conversion. Im inclined to leave this as is to avoid doing something that adds complexity or goes against the standards.

Comment (jonathanlew) [reply src/software/backend/wifi_backend.cpp:53]: I think we should either use unique_ptrs everywhere or use them nowhere. We can discuss this offline if you'd like

File: src/software/backend/wifi_backend.h
Comment (MathewMacDougall) [reply src/software/backend/wifi_backend.h:13]: Ah this is the same problem as Plays (and the factory). IIRC we considered
* static class functions that return the correct subconfig from some top-level config
* Manually registering things in factories rather than using the generic one

Haven't put a ton of thought into this yet.

File: src/shared/constants.h
Comment (kelvinkoon) [inline src/shared/constants.h:91]: Definitely preferring how these are defines now. :+1:

File: src/software/parameter/config/network.yaml
Comment (kelvinkoon) [inline src/software/parameter/config/network.yaml]: Typo on `thunderbots` presumably.

**Review by MathewMacDougall — COMMENTED** [review]
Nothing to add beyond existing comments atm.

File: src/software/proto/message_translation/protobuf_message_translator.h
Comment (akhilveeraghanta) [reply src/software/proto/message_translation/protobuf_message_translator.h]: removed after offline discussion

File: src/software/parameter/config/network.yaml
Comment (akhilveeraghanta) [reply src/software/parameter/config/network.yaml]: fixed

File: src/shared/constants.h
Comment (akhilveeraghanta) [reply src/shared/constants.h:91]: 💯

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Review by kelvinkoon — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]

**Pull request merged**
